File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.cpp
class Solution {
public:
    int numberOfSpecialChars(string word) {
        // TC : O(n + 26) ≈ O(n)
        // SC : O(1)
        vector<int> lastlower(26, -1);
        vector<int> firstupper(26, -1);

        for (int i = 0; i < word.size(); i++) {
            char c = word[i];
            if (c >= 'a' && c <= 'z') {
                lastlower[word[i] - 'a'] = i;
            } else if (c >= 'A' && c <= 'Z') {
                if (firstupper[word[i] - 'A'] == -1) {
                    firstupper[word[i] - 'A'] = i;
                }
            }
        }

        int countSpecialChars = 0;
        for (int i = 0; i < 26; i++) {
            if (lastlower[i] != -1 && firstupper[i] != -1 &&
                lastlower[i] < firstupper[i]) {
                countSpecialChars++;
            }
        }
        return countSpecialChars;
    }
};

File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.java
class Solution {
    public int numberOfSpecialChars(String word) {
        // TC : O(n + 26) ≈ O(n)
        // SC : O(1)
        int [] lastLower = new int[26];
        int [] firstUpper = new int[26]; 

        for(int i  =0  ; i< 26 ; i++) {
            lastLower[i] = -1;
            firstUpper[i] = -1;
        }

        for(int i = 0 ; i< word.length(); i++) {
            char ch = word.charAt(i);
            if(Character.isLowerCase(ch)) {
                lastLower[ch-'a'] = i;
            }
            else if(Character.isUpperCase(ch)) {
                if(firstUpper[ch-'A'] == -1) {
                    firstUpper[ch - 'A'] = i;
                }
            }
        }

        int countChar = 0;
        for(int i = 0 ; i < 26; i++) {
            if(lastLower[i] != -1 && firstUpper[i] != -1 && lastLower[i] < firstUpper[i] ) {
                countChar++;
            }
        }

        return countChar;
    }
}